[sdl3-shadercross] add new port

Agorath opened the pull request:
This pull request adds a new port with a preview version of the upcoming SDL_shadercross library.

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

Agorath committed:
[sdl-shadercross] add new port

Agorath committed:
Update versions database.

Agorath committed:
Added BISON as it is required for non windows targets.

dg0yt commented on ports/sdl-shadercross/portfile.cmake:
```diff
@@ -0,0 +1,42 @@
+    HEAD_REF main
+)
+
+if(NOT VCPKG_TARGET_IS_WINDOWS)
+    vcpkg_find_acquire_program(BISON)
+    get_filename_component(BISON_DIR "${BISON}" DIRECTORY)
+    vcpkg_add_to_path(PREPEND "${BISON_DIR}")
+endif()
```
Isn't bison required for non-Windows?

Instead of messing with PATH, can the executable be passed as a parameter in `vcpkg_cmake_configure`?

Agorath replied:
> Isn't bison required for non-Windows?

Yes, that's correct. Is `if(NOT VCPKG_TARGET_IS_WINDOWS)` incorrect in only getting Bison for non-Windows, or do you mean it might also be required for MinGW, which should thus be excluded via `AND NOT VCPKG_TARGET_IS_MINGW`?
> Instead of messing with PATH, can the executable be passed as a parameter in `vcpkg_cmake_configure`?

I only stumbled upon the Bison requirement for non-windows when creating this pull request, as I currently code in Windows exclusively. I then looked up how to get Bison, looked at what other ports are doing to achieve that and used that code. I'll try adding it as a parameter via `vcpkg_cmake_configure` when I look at what's causing the problem with the x64-Linux build later today.

Agorath replied:
If I'm not missing anything, there doesn't seem to be a way to add BISON via the `vcpkg_cmake_configure` call. There's no BISON related variables in the CMakeLists.txt of the package as it looks for it via `find_package`:
![image](https://github.com/user-attachments/assets/756428b7-d71c-4165-8416-01437a2fbc8e)

dg0yt commented on ports/sdl-shadercross/portfile.cmake:
```diff
@@ -0,0 +1,42 @@
+if(VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW)
+    set(config_path "share/${PORT}")
+endif()
+vcpkg_cmake_config_fixup(CONFIG_PATH "${config_path}")
```
```suggestion
vcpkg_cmake_config_fixup(CONFIG_PATH "${config_path}" PACKAGE_NAME sdl3_shadercross)
```
for the underscore in the CMake package name?

Agorath replied:
Good catch. I missed that. After giving it some consideration, I think all the `config_path` shenanigans above are unnecessary, and the following should be enough. What do you think about that?
```suggestion
vcpkg_cmake_config_fixup(PACKAGE_NAME "sdl3_shadercross" CONFIG_PATH "share/${PORT}")
```

dg0yt replied:
I suggest to use the lower-case PACKAGE_NAME. It is actually only the name of the directory created in `share/${PORT}`, and (there was a time when) vcpkg was preferring lower-case. (Nitpick.)

Agorath commented on ports/sdl-shadercross/portfile.cmake:
```diff
@@ -0,0 +1,42 @@
+
+vcpkg_cmake_install()
+
+set(config_path "share/${PORT}/SDL3_shadercross")
+if(VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW)
+    set(config_path "share/${PORT}")
+endif()
+vcpkg_cmake_config_fixup(CONFIG_PATH "${config_path}")
```
The following is what I meant in my previous comment.
I think this should work, as the correct install directory is set for every target platform in `vcpkg_cmake_configure` via `-DSDLSHADERCROSS_INSTALL_CMAKEDIR_ROOT=share/${PORT}`
```suggestion
vcpkg_cmake_config_fixup(PACKAGE_NAME "SDL3_shadercross" CONFIG_PATH "share/${PORT}")
```

Agorath committed:
Merge commit 'efb1e7436979a30c4d3e5ab2375fd8e2e461d541' into sdl-shadercross

# Conflicts:
#	versions/baseline.json

Agorath committed:
Renamed the new port from sdl-shadercross to sdl3-shadercross.

Fixed some issues in portfile.cmake.

Agorath committed:
Merge branch 'microsoft:master' into sdl-shadercross

Agorath committed:
Alphabetic reordering

Agorath commented:
I haven't yet been able to figure out what the exact problem with the Linux build is. I set up a Linux machine and tried debugging the build, but haven't been able to figure it out, yet. But I'm on it.
I'd like to propose the following:
I'll set the port in its current iteration as incompatible with Linux and mark it ready for review. This way it could be merged in case there are no issues or objections.
In parallel, I'll continue working on finding out how to fix the Linux build issues and possibly fixing those with a new port-version?

Would that be a viable way forward?

Agorath committed:
Added linux to the unsupported platforms for now.

Agorath committed:
Facepalm for forgetting to change & to | after copy&pasting...

dg0yt commented:
> I haven't yet been able to figure out what the exact problem with the Linux build is. I set up a Linux machine and tried debugging the build, but haven't been able to figure it out, yet.

CI log:
~~~
CMake Error at /vcpkg/downloads/tools/cmake-3.30.1-linux/cmake-3.30.1-linux-x86_64/share/cmake-3.30/Modules/FindPackageHandleStandardArgs.cmake:233 (message):
  Could NOT find DirectXShaderCompiler (missing:
  DirectXShaderCompiler_INCLUDE_PATH)
Call Stack (most recent call first):
  /vcpkg/downloads/tools/cmake-3.30.1-linux/cmake-3.30.1-linux-x86_64/share/cmake-3.30/Modules/FindPackageHandleStandardArgs.cmake:603 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindDirectXShaderCompiler.cmake:37 (find_package_handle_standard_args)
  /vcpkg/scripts/buildsystems/vcpkg.cmake:893 (_find_package)
  CMakeLists.txt:154 (find_package)
~~~

cmake/FindDirectXShaderCompiler.cmake, non-WIN32:
~~~
find_path(DirectXShaderCompiler_INCLUDE_PATH NAMES "dxcapi.h" PATH_SUFFIXES "include" "include/dxc" "linux/include" "linux/include/dxc")
~~~

installed:
~~~
directx-dxc:x64-linux:/include/directx-dxc/WinAdapter.h
directx-dxc:x64-linux:/include/directx-dxc/dxcapi.h
directx-dxc:x64-linux:/include/directx-dxc/dxcerrors.h
directx-dxc:x64-linux:/include/directx-dxc/dxcisense.h
~~~~

So maybe try with 
~~~
   OPTIONS
       "-DDirectXShaderCompiler_INCLUDE_PATH=${CURRENT_INSTALLED_DIR}/include/directx-dxc"
~~~

(Does it really pick the right paths for Windows?)

dg0yt commented on ports/sdl3-shadercross/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+    HEAD_REF main
+)
+
+if(NOT VCPKG_TARGET_IS_WINDOWS)
+    vcpkg_find_acquire_program(BISON)
+    get_filename_component(BISON_DIR "${BISON}" DIRECTORY)
+    vcpkg_add_to_path("${BISON_DIR}")
+endif()
+
```
This seems to be unused with `-DSDLSHADERCROSS_INSTALL_RUNTIME=OFF`.
```suggestion
```

Agorath replied:
Good catch, I sadly missed that.

dg0yt commented on ports/sdl3-shadercross/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        -DSDLSHADERCROSS_INSTALL=ON
+        -DSDLSHADERCROSS_INSTALL_CMAKEDIR_ROOT=share/${PORT}
```
```suggestion
        -DSDLSHADERCROSS_INSTALL_CMAKEDIR_ROOT=share/sdl3_shadercross
```

dg0yt commented on ports/sdl3-shadercross/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+vcpkg_copy_pdbs()
+vcpkg_fixup_pkgconfig()
+
+vcpkg_cmake_config_fixup(CONFIG_PATH share/${PORT} PACKAGE_NAME "sdl3_shadercross")
```
```suggestion
vcpkg_cmake_config_fixup(PACKAGE_NAME "sdl3_shadercross")
```
as a result of the other change.

dg0yt commented on ports/sdl3-shadercross/usage:
```diff
@@ -0,0 +1,4 @@
+sdl3-shadercross provides CMake targets:
+
+  find_package(SDL3_shadercross CONFIG REQUIRED)
+  target_link_libraries(main PRIVATE SDL3_shadercross::SDL3_shadercross)
```
```suggestion
  target_link_libraries(main PRIVATE SDL3_shadercross::SDL3_shadercross)
```
Please ensure a line break at the end of the line. (It is a new empty line in VS Code...)

Agorath replied:
I added the line break.

Agorath committed:
Review changes.

Agorath committed:
Reactivated the Linux platform.

Agorath commented:
> So maybe try with
> 
> ```
>    OPTIONS
>        "-DDirectXShaderCompiler_INCLUDE_PATH=${CURRENT_INSTALLED_DIR}/include/directx-dxc"
> ```
> 
> (Does it really pick the right paths for Windows?)

Thanks for the feedback, you are indeed correct!
I just had a look at `config-x64-windows-rel-CMakeCache.txt.log` on my Windows machine, and it indeed didn't use the vcpkg `directx-dxc` version, but instead found the one from the Windows SDK and used that instead:
`DirectXShaderCompiler_INCLUDE_PATH:PATH=C:/Program Files (x86)/Windows Kits/10/Include/10.0.26100.0/um`

With the proposed change, the vcpkg port of directx-dxc is now correctly found on Windows:
DirectXShaderCompiler_INCLUDE_PATH:PATH=D:/Coding/git_repos/SDL3_GPU_Experiments/vcpkg_packages/x64-windows/include/directx-dxc

Agorath committed:
Updated to the latest commit of the SDL_shadercross repo.

dg0yt commented:
Do downstream users need the headers from DirectXShaderCompiler?

If yes, we are in trouble again: The installed CMake config has
~~~
find_dependency(DirectXShaderCompiler)
~~~
You might need to patch the find module. Or make directx-dxc put the headers to an expected location?

Agorath commented:
> Do downstream users need the headers from DirectXShaderCompiler?
> 
> If yes, we are in trouble again: The installed CMake config has
> 
> ```
> find_dependency(DirectXShaderCompiler)
> ```
> 
> You might need to patch the find module. Or make directx-dxc put the headers to an expected location?

I just had a look at the code, and I don't think so.
The only thing that's required to use SDL3_ShaderCross is to include `#include <SDL3_shadercross/SDL_shadercross.h>`. That header itself only includes SDL3 headers and declares the functions SDL3_ShaderCross provides, and that's it. The rest is compiled into the shared library.

Agorath commented:
I just perused the code of SDL3_Shadercross, and they don't include any of the headers of the directx-dxc library at all. That's also the reason there are no conflicts with the differing include path being used in the directx-dxc port.
They seem to use the lib by forward declaring the functions they utilise from the lib and letting the linker handle the rest.
![image](https://github.com/user-attachments/assets/a751c641-d7be-4f79-93ab-749ded4d0fc5)

BillyONeal commented:
Can you fix it to not need the `find_dependency` calls?

Agorath committed:
Merge branch 'microsoft:master' into sdl-shadercross

Agorath committed:
Merge branch 'microsoft:master' into sdl-shadercross

Agorath committed:
Updated to the latest commit of the SDL_shadercross repo.

Agorath committed:
Updated the ports' dependencies to the latest respective versions.

Agorath committed:
DirectXShaderCompiler include search removed.

This should, hopefully, finally fix the build problems caused by not finding the non-needed DXC headers.

Agorath committed:
usage was in the wrong directory.

Agorath committed:
Fix for Windows compilation.

BillyONeal reviewed (approved):
Thanks for the new port!
